tiledb/api/src/array/dimension/mod.rs

```diff
@@ -207,62 +209,63 @@ impl<'ctx> Builder<'ctx> {
         })
     }
 
-    /// N.B., This API will eventually be removed when we fixup the current
-    /// implementation of `new` to take a `DimensionConstraint` argument instead
-    /// of the serde values for range and extent.
-    pub fn new_string(
+    pub fn new_optional(
```

Comment:
I'm not a super huge fan of this name, but I couldn't come up with anything better.

Reply:
There's no guidance on which to use, `new` or `new_optional`. Perhaps writing it would inspire the right name.

Reply:
I think I still like my suggestion to have one function the most with something like
```
pub enum DimensionDatatype {
    UInt8 { domain: std::ops::Range<u8>, extent: Option<u8> },
    UInt64 { domain, std::ops::Range<u64>, extent: Option<u64> },
    ...,
    StringAscii,
    StringUtf8
}
```

Reply:
I'm mulling over redoing the whole Builder/Constructor pattern. One thought I had to clean this up quite a bit would be to switch this Builder to be more like the Enumeration version where the only FFI calls happen in `build()` and everything else is just setting values on a DimensionData instance. Although, even then I'm not sure how to make new() work with and without a Range that isn't super misleading to users. I'll keep thinking on it.

Reply:
Yeah, I might go that way. I started to go that route originally but it felt weird so I didn't follow through on it. Though now that I've got the current shape of things I'll give it another whirl to see if clears up the issues with `new`.

Reply:
Any reason you went with std::ops::Range over `[T; 2]`? The array approach means that Vec allocation goes away which would be nice.

Reply:
You're right, `[T; 2]` should be better for that.  It is nice that range carries the semantic meaning though.

I guess you could construct `[T; 2]` trivially from a range, so you wouldn't need Vec even with the range.

I don't really have a preference

Reply:
> I'm mulling over redoing the whole Builder/Constructor pattern. One thought I had to clean this up quite a bit would be to switch this Builder to be more like the Enumeration version where the only FFI calls happen in build() and everything else is just setting values on a DimensionData instance

I'm not sure how well that solves this particular problem but I have had this thought that this is something we could apply more broadly.  It would be kind of nice to get the `TileDBResult` out of so many of the setter functions

Reply:
Yeah, I think the DimensionDatatype will likely end up being cleaner for the API rather than whatever I was thinking. I'm gonna go start my caffeination protocols and will try that approach.

Totes agree that those Result's going away would be nice. I've also noticed that the using the `impl Factor for $ThingData` is quite lovely for a more programatic API interaction.